File: in_Java/0116 Populating Next Right Pointers in Each Node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null) return root;
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        Node cur;
        Node pre;
        int n;
        while(!q.isEmpty()){
            n = q.size();
            pre = q.poll();
            if(pre.left != null) q.add(pre.left);
            if(pre.right != null) q.add(pre.right);
            n--;
            while(n-- != 0){
                cur = q.poll();
                pre.next = cur;
                if(cur.left != null) q.add(cur.left);
                if(cur.right != null) q.add(cur.right);
                pre = cur;
            }
            pre.next = null;
        }
        return root;
    }
}

File: in_Python/0116 Populating Next Right Pointers in Each Node.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if root == None:
            return root
        q = [root]
        while q:
            cur = q.pop(0)
            if cur.left and cur.right:
                cur.left.next = cur.right
                if cur.next:
                    cur.right.next = cur.next.left
                q.append(cur.left)
                q.append(cur.right)
        return root